### ISSUE TITLE
Removing a custom resource can have inconsistencies

### ISSUE THREAD

**Issue by mwtian** [ray-project/ray#16696]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray at HEAD, Python 3.7.

### Reproduction (REQUIRED)
`python releaser/e2e.py --test-config ray/release/xgboost_tests/xgboost_tests.yaml --test-name train_small`
- Some actors fail to schedule even after nodes with enough resources are available.

There can be other issues causing actors failing to schedule, but one confusion I have is the logic of removing a resource. When a resource is available on multiple nodes, removing the resource the first time from a node would delete the resource ID and name from [string_to_int_map_](https://github.com/ray-project/ray/blob/54ce8092abdac461167ca064092ba330b9625cc7/src/ray/raylet/scheduling/cluster_resource_scheduler.cc#L556). If the logic attempts to delete the same resource name from another node, because the resource name is no longer available in `string_to_int_map_`, the deletion becomes a no-op. If the resource name is added back to `string_to_int_map_` again, the resource will suddenly become available in the node that should have deleted the resource.

I have seen placement groups can add and delete the same resource name across multiple nodes, which should trigger this issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by DmitriGekhtman** [comment]
cc @rkooo567

**Comment by wuisawesome** [comment]
cc @lixin-wei 

If you're able to make the cluster resource scheduler use a `std::unordered_map<std::string, FixedPoint>` in your cleanup, you will fix this issue.

**Comment by lixin-wei** [comment]
> When a resource is available on multiple nodes, removing the resource the first time from a node would delete the resource ID and name from string_to_int_map_. If the logic attempts to delete the same resource name from another node, because the resource name is no longer available in string_to_int_map_, the deletion becomes a no-op.

I don't understand this. Every node has its own string_to_int_map_, why removing a resource from a node will affect another node?

**Comment by lixin-wei** [comment]
@wuisawesome We shouldn't use string keys, this will be a huge performence regression. To fix this issue I think we can keep the hash result, don't remove items from string_to_int_map_ when removing resources.

**Comment by mwtian** [comment]
> @wuisawesome We shouldn't use string keys, this will be a huge performence regression. To fix this issue I think we can keep the hash result, don't remove items from string_to_int_map_ when removing resources.

Not removing the items are great. If we want to avoid the complexity of passing `string_to_int_map_` for translation while maintaining performance, we can intern the string and use pointers to interned objects for comparisons. Maybe boost::flyweight can be used but I haven't verified.

**Comment by lixin-wei** [comment]
@mwtian Did you mean `std::unordered_map<boost::flyweight<std::string>, FixedPoint>` ? I think it's a good idea.

**Comment by lixin-wei** [comment]
@mwtian Will you work on this? I think it worth the effort, we can set a task to the milestone. If you don't have time I can help.
@wuisawesome Do you agree with replacing it with `std::unordered_map<boost::flyweight<std::string>, FixedPoint>` ?

**Comment by mwtian** [comment]
> @mwtian Did you mean `std::unordered_map<boost::flyweight<std::string>, FixedPoint>` ? I think it's a good idea.
> Will you work on this? I think it worth the effort, we can set a task to the milestone. If you don't have time I can help.

Sent #16888 to drop the `Remove()` method. We can convert map keys to `boost::flyweight<std::string>` in a subsequent refactoring.

**Comment by rkooo567** [comment]
Is it closeable?

**Comment by mwtian** [comment]
Let's close this for now.
